runtime/memprof.c

```diff
@@ -346,18 +318,20 @@ void caml_memprof_handle_postponed(void)
   }
 
   while (postponed_tl != postponed_hd) {
-    struct postponed_block pb = *postponed_tl;
-    block = pb.block;           /* pb.block is not a root! */
+    caml_callstack* callstack = &postponed_tl->callstack;
+    uintnat occurrences = postponed_tl->occurrences;
+    enum ml_alloc_kind kind = postponed_tl->kind;
+    block = postponed_tl->block;
     postponed_pop();
-    if (postponed_tl == postponed_hd) purge_postponed_queue();
 
     /* If using threads, this call can trigger reentrant calls to
        [caml_memprof_handle_postponed] even though we set
        [caml_memprof_suspended]. */
     ephe = do_callback(Tag_val(block), Wosize_val(block),
-                       pb.occurrences, pb.callstack, pb.kind);
+                       occurrences, callstack, kind);
 
     if (Is_block(ephe)) caml_ephemeron_set_key(Field(ephe, 0), 0, block);
+    if (postponed_tl == postponed_hd) purge_postponed_queue();
```

Comment:
Also, I tried to maintain the invariant that, when OCaml is executed, the queue is purged as soon as it is empty. This breaks the invariant, since the queue is purged after the callback is called.